### PR TITLE
Fix Product Switch new Amount for the Upgrade Support journey

### PR DESCRIPTION
**Pull request by andresilva-guardian**
## Context
Associated manage frontend pr: https://github.com/guardian/manage-frontend/pull/1545

This PR addresses Bug Report 328 - a Severity 3 (Urgent) issue reported via the Guardian's customer support system. The bug was identified when a customer tried to switch their recurring contribution to Supporter Plus with a monthly payment of €20.00, but the system only set it to €12.00 instead.

**Trello Card**: [Amount doesn't update correctly on product switch](https://trello.com/c/jpf669xe/908-amount-doesnt-update-correctly-on-product-switch-severity-3-urgent-bug-report-328-system-impacted-manage-my-account-mma)

**Customer Impact Details**:
- **Affected System**: Manage My Account (MMA)
- **Customer Journey**: Billing (Payments, Payment Methods, Refunds) - S+ amount increase process
- **Impact Severity**: 3 - Urgent
- **Users Impacted**: Single user reported, but affects all users attempting to switch from recurring contributions to Supporter Plus with custom amounts

## What does this change?

**Problem**: The product-switch API for `recurring-contribution-to-supporter-plus` was ignoring user-selected amounts when switching from recurring contributions to Supporter Plus subscriptions. The API only accepted `preview` and `applyDiscountIfAvailable` parameters but had no mechanism to handle custom amounts selected by users in the frontend.

**Root Cause Analysis**: 
- The `getSwitchInformationWithOwnerCheck` function in `switchInformation.ts` was calculating contribution amounts based only on the user's previous contribution amount (`previousAmount`)
- The API schema in `schemas.ts` didn't include a field for user-desired amounts
- This resulted in the system using the previous amount or falling back to minimum Supporter Plus pricing instead of the user's selected amount

**Solution**: 
1. **Extended API Schema**: Added an optional `newAmount` parameter to `productSwitchRequestSchema` to accept user-desired amounts
2. **Updated Calculation Logic**: Modified the contribution calculation to use `userDesiredAmount` (from `input.newAmount` or fallback to `previousAmount`)
3. **Added Validation**: Implemented validation to ensure the desired amount meets minimum Supporter Plus pricing requirements when explicitly provided
4. **Backward Compatibility**: Maintained existing behavior when `newAmount` is not provided to avoid breaking existing integrations

**Key Changes**:
- **Schema Extension** (`schemas.ts`): Added `newAmount: z.optional(z.number().positive())` to support user-selected amounts
- **Business Logic** (`switchInformation.ts`): 
  - Calculate contribution based on `userDesiredAmount` instead of `previousAmount`
  - Validate minimum amount requirements when `newAmount` is explicitly provided
  - Use `userDesiredAmount` as the `actualTotalPrice` to reflect the user's selection
- **Test Coverage**: Added comprehensive tests covering new amount scenarios, backward compatibility, and validation edge cases

## How to test

- Run the existing test suite: `npm test` in the `handlers/product-switch-api` directory
- New test cases specifically cover:
  - Switch with custom `newAmount` parameter
  - Backward compatibility without `newAmount` 
  - Validation error when `newAmount` is below minimum Supporter Plus price

## How can we measure success?

**Key Success Metrics**:

1. **Bug Resolution**: Zero reports of amount discrepancies in product switches from recurring contributions to Supporter Plus
2. **Customer Support**: Reduction in escalated cases related to incorrect Supporter Plus amounts after product switches
3. **Revenue Accuracy**: Supporter Plus subscriptions reflect the exact amounts selected by users during the switch process
4. **User Experience**: Users can successfully upgrade to Supporter Plus with their desired contribution amounts without manual intervention from customer support

**Monitoring Recommendations**:
- Filter logs for `product-switch-api` calls with `newAmount` parameter to track adoption
- Set up alerts for validation errors to identify potential UX issues with amount selection
- Monitor customer support tickets tagged with subscription amount discrepancies

## Have we considered potential risks?

**Identified Risks & Mitigations**:

1. **Backward Compatibility Risk**: 
   - **Risk**: Existing integrations might break if they depend on current calculation logic
   - **Mitigation**: The `newAmount` parameter is optional, maintaining existing behavior when not provided

2. **Revenue Impact Risk**:
   - **Risk**: Users might select amounts lower than intended business thresholds
   - **Mitigation**: Validation ensures amounts meet minimum Supporter Plus pricing requirements

3. **Data Validation Risk**:
   - **Risk**: Invalid or malicious amount values could cause system errors
   - **Mitigation**: Zod schema validation ensures positive numbers only

4. **Customer Experience Risk**:
   - **Risk**: Validation errors might confuse users when they select invalid amounts
   - **Mitigation**: Clear error messages guide users toward appropriate amounts and suggest alternative APIs

**Backlog Items**:
- Frontend updates to pass `newAmount` parameter when users select custom amounts

## Images

N/A - This is a backend API change with no direct UI modifications.

## Accessibility

N/A - This change is backend-only and doesn't affect any user interface elements.

### CODE REVIEW

File: handlers/product-switch-api/src/switchInformation.ts [hunk 1]
```diff
@@ -192,7 +192,17 @@ export const getSwitchInformationWithOwnerCheck = async (
 	const actualBasePrice =
 		maybeDiscount?.discountedPrice ?? catalogInformation.supporterPlus.price;
 
-	const contributionAmount = Math.max(0, previousAmount - actualBasePrice);
+	const userDesiredAmount = input.newAmount ?? previousAmount;
```
Comment (johnduffell) [inline handlers/product-switch-api/src/switchInformation.ts:195]:
```suggestion
  // newAmount is only passed in where the user is in the switch journey - for cancellation saves they always get the base price (with discount)
	const userDesiredAmount = input.newAmount ?? previousAmount;
```

Comment (johnduffell) [reply handlers/product-switch-api/src/switchInformation.ts:195]: something like that might help future readers! it would also be worth validating that newamount is set for switches, and is not set for saves, just so we can recheck this logic if something changes.

File: handlers/product-switch-api/src/switchInformation.ts [hunk 2]
```diff
@@ -192,7 +192,17 @@ export const getSwitchInformationWithOwnerCheck = async (
...
+	
+	// Validate that the user's desired amount is at least the base Supporter Plus price
+	// Only validate when newAmount is explicitly provided by the frontend
+	if (input.newAmount && userDesiredAmount < actualBasePrice) {
+		throw new ValidationError(
+			`Cannot switch to Supporter Plus: desired amount (${userDesiredAmount}) is less than the minimum Supporter Plus price (${actualBasePrice}). Use the update-supporter-plus-amount API to modify contribution amounts instead.`
```
Comment (johnduffell) [inline handlers/product-switch-api/src/switchInformation.ts:201]: great error message - however it would actually be members data api to change contribution amounts

**Review by johnduffell — APPROVED** [review]
I'll approve as none of the comments/suggestions I made would affect the fact that it fixes the PROD bug.  They can be fixed in a followup.

The PR description is very long so I didn't actually read it all, hopefully there was nothing essential further down.
